### PR TITLE
More support for classical simulation

### PR DESCRIPTION
**Pull request by mpharrigan**
- use idiomatic `on_classical_val` methods in arithmetic
 - support some more uncomputations

### CODE REVIEW

**Comment by mpharrigan** [comment]
This contains the non-superfluous changes that originated from #575

**Review by fdmalone — DISMISSED** [review]

**Comment by mpharrigan** [comment]
ah, right. Now I know why I expanded the scope in #575. If `cirq.CX` is used, we can't support classical simulation.

**Comment by fdmalone** [comment]
What's the status with this one? I have some vested interested in the interception of the cirq basic gates in favor of bloqs :)

Screenshot 2024-01-26 at 2 20 48 PM

**Comment by mpharrigan** [comment]
I think the quote-unquote right way forward is to:

 1. make sure we have bloqs in this library for everything including CZ
 2. following #591 , we can replace instances of `cirq.CNOT` with our CNOT bloq 

We'd do step 2 for all cases where we need to have control over the protocols: e.g. in this PR I want to add classical simulation to CNOT which I can't do to `cirq.CNOT` and you want to add wire symbols to `cirq.CZ`, which honestly should work via the cirq interop and I don't know why it's not translating the `@` but nonetheless we should turn it into a CZ bloq

**Comment by mpharrigan** [comment]
I changed `Add` to a `Bloq` and used the `CNOT` bloq in place of `cirq.CNOT`, which unblocked the classical test. I did light modifications to the other `Add` tests to use bloq idioms.

File: qualtran/bloqs/arithmetic/addition.py [hunk 1]
```diff
@@ -69,15 +73,26 @@ class Add(GateWithRegisters, cirq.ArithmeticGate):
...
+        incoming: Dict[str, 'SoquetT'],
+        outgoing: Dict[str, 'SoquetT'],
+    ):
+        import quimb.tensor as qtn
+
+        N = 2**self.bitsize
+        inds = (incoming['a'], incoming['b'], outgoing['a'], outgoing['b'])
+        unitary = np.zeros((N,) * len(inds), dtype=np.complex128)
+        for a, b in itertools.product(range(N), range(N)):
+            unitary[a, b, a, (a + b) % N] = 1
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/addition.py:90]: Why is this % N

Comment (mpharrigan) [reply qualtran/bloqs/arithmetic/addition.py:90]: otherwise it would be out of range

**Comment by mpharrigan** [comment]
Tests failing because `Add(bitsize=1)` used to work ...

File: qualtran/bloqs/arithmetic/addition.py (hunk 1 above)
Comment (fdmalone) [reply qualtran/bloqs/arithmetic/addition.py:95]: Do you know what happens with overflow? I'm sort of curious.

Comment (mpharrigan) [reply qualtran/bloqs/arithmetic/addition.py:95]: The tests are written to never trigger overflow, which is something to say. 

looking at `cirq.ArithmeticGate` line 219 seems to imply that `% N` is used, which was the previous behavior.

**Comment by mpharrigan** [comment]
I factored out #623

**Pull request merged**